Objects/typeobject.c

```diff
@@ -10389,9 +10389,13 @@ supercheck(PyTypeObject *type, PyObject *obj)
         Py_XDECREF(class_attr);
     }
 
-    PyErr_SetString(PyExc_TypeError,
+    PyErr_Format(PyExc_TypeError,
                     "super(type, obj): "
-                    "obj must be an instance or subtype of type");
+                    "obj must be an instance or subtype of type. "
+                    "Got obj: '%.200s', type: '%.200s'.",
```

Comment:
1. I think it will read more naturally if we put the extra information inline in parentheses, rather than tacking it on at the end.
2. I think it could also be important for the user to be able to distinguish whether `obj` was an instance or a type?
3. If we are giving the actual types found, I think it's more natural to phrase the error in terms of what we found to (not) be the case, rather than in terms of what must be the case.

Putting those comments together, my preference would be for an error message in one of these two forms:

> super(type, obj): obj (instance of FooType) is not an instance or subtype of type (BarType)

or, if a type was passed in for `obj`:

> super(type, obj): obj (FooType) is not an instance or subtype of type (BarType)

What do you think?

cc @pablogsal who cares about useful error messages.

Reply:
I like that. However, if we go this way with verbose error message I would suggest slightly modifying your second error message example to be:
`super(type, obj): obj (type FooType) is not an instance or subtype of type (BarType)`

to mention explicitly that obj is type itself.